File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    private static int[] dist = {1, 0, -1, 0, 1};
    
    private static class Point{
        int x, y;
        public Point(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        int[][] board = new int[maps.length][maps[0].length];
        ArrayDeque<Point> queue = new ArrayDeque<>();
        
        queue.add(new Point(0, 0));
        
        board[0][0] = 1;
        board[maps.length-1][maps[0].length-1] = -1;
        
        while(!queue.isEmpty()){
            Point now = queue.poll();
            
            for(int i = 0; i < 4; i++){
                int nr = now.x + dist[i];
                int nc = now.y + dist[i+1];
                if(nr < 0 || nc < 0|| nr >= maps.length || nc >= maps[0].length || board[nr][nc] > 0 || maps[nr][nc] == 0){
                    continue;
                }
                queue.add(new Point(nr, nc));
                board[nr][nc] = board[now.x][now.y] + 1;
            }
        }
        
        
        
        return board[maps.length-1][maps[0].length-1];
    }
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    
    n = len(maps)
    m = len(maps[0])
    
    board = [[0]*m for _ in range(n)]
    
    board[0][0] = 1
    board[n-1][m-1] = -1
    
    queue = deque([(0, 0)])
    
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    
    while queue:
        row, col = queue.popleft()
        for i in range(4):
            nr = row + dy[i]
            nc = col + dx[i]
            if nr < 0 or nc < 0 or nr >= n or nc >= m:
                continue
            if maps[nr][nc] == 0:
                continue
            if board[nr][nc] > 0:
                continue
            
            board[nr][nc] = board[row][col] + 1
            queue.append((nr, nc))
            
    return board[n-1][m-1]